mixer/adapter/svcctrl/quotaprocessor.go

```diff
@@ -0,0 +1,220 @@
+	quotaIndex map[string]*config.Quota
+	client     serviceControlClient
+}
+
+func getQuotaMode(args adapter.QuotaArgs) string {
+	if args.BestEffort {
+		return quotaModeBestEffort
+	}
+	return quotaModeNormal
+}
+
+func createQuotaResult(status rpc.Status, expiration time.Duration, amount int64) adapter.QuotaResult {
+	return adapter.QuotaResult{
+		Status:        status,
+		ValidDuration: expiration,
+		Amount:        amount,
+	}
+}
+
+func quotaResultFromQuotaError(quotaErr *sc.QuotaError, quotaCfg *config.Quota) adapter.QuotaResult {
+	code := serviceControlErrorToRPCCode(quotaErr.Code)
+	return createQuotaResult(status.WithMessage(code, quotaErr.Description),
+		toDuration(quotaCfg.Expiration), 0)
+}
+
+func makeQuotaResult(response *sc.AllocateQuotaResponse, quotaCfg *config.Quota,
+	args *adapter.QuotaArgs) adapter.QuotaResult {
+	succeeded := response.AllocateErrors == nil || len(response.AllocateErrors) == 0
+	if !succeeded {
+		return quotaResultFromQuotaError(response.AllocateErrors[0], quotaCfg)
+	}
+
+	// If quota metric is not found in response, just return the request quota amount
+	if response.QuotaMetrics == nil || len(response.QuotaMetrics) == 0 {
+		return createQuotaResult(status.OK, toDuration(quotaCfg.Expiration), args.QuotaAmount)
+	}
+
+	return createQuotaResult(status.OK, toDuration(quotaCfg.Expiration),
+		getAllocatedQuotaAmount(args, response.QuotaMetrics, quotaCfg))
+}
+
+func buildAllocateRequest(instance *quota.Instance,
+	args adapter.QuotaArgs, quotaCfg *config.Quota) (*sc.AllocateQuotaRequest, error) {
+	rawAPIKey, ok := dimensionToString(instance.Dimensions, apiKeyDimension)
+	if !ok {
+		return nil, fmt.Errorf("dimension %v not found in instance %v", apiKeyDimension, *instance)
+	}
+
+	apiOperation, ok := dimensionToString(instance.Dimensions, apiOperationDimension)
+	if !ok {
+		return nil, fmt.Errorf("dimension %v not found in instance %v", apiOperationDimension, *instance)
+	}
+
+	// If GoogleQuotaMetric name is not set in config, assume it is the same of Istio quota name.
+	metricName := quotaCfg.GoogleQuotaMetricName
+	if metricName == "" {
+		metricName = quotaCfg.Name
+	}
+	apiKey := generateConsumerIDFromAPIKey(rawAPIKey)
+	op := &sc.QuotaOperation{
+		ConsumerId:  apiKey,
+		MethodName:  apiOperation,
+		OperationId: uuid.New(),
+		QuotaMetrics: []*sc.MetricValueSet{
+			{
+				MetricName: metricName,
+				MetricValues: []*sc.MetricValue{
+					{
+						Int64Value: getInt64Address(args.QuotaAmount),
+					},
+				},
+			},
+		},
+		QuotaMode: getQuotaMode(args),
+	}
+
+	return &sc.AllocateQuotaRequest{
+		AllocateOperation: op,
+	}, nil
+}
+
+func getAllocatedQuotaAmount(args *adapter.QuotaArgs, quotaMetrics []*sc.MetricValueSet,
+	quotaCfg *config.Quota) int64 {
+	// Initialize allocated amount to requested amount
+	var amount = args.QuotaAmount
+	for _, metric := range quotaMetrics {
+		if metric.MetricName == quotaAllocationResultMetric &&
```

Comment:
Is it worth to build a reverse map from Google quota name to mixer quota name?

Reply:
Since today, we handle one quota at a time, it doens't help improve efficiency.